### PR TITLE
[4.0] Workflow stage delete publish

### PR DESCRIPTION
**Pull request by chmst**
Pull Request for Issue #21846 

### Summary of Changes
Remove the method delete() from stageTable to stageModel.
Doing this some fixes have been added, comments and code have been improved

### Testing Instructions
Make a workflow with several stages and transitions.
Now try to change the status of one or more stages, thrash them, unpublish them.

If you try to trash a default state, you must get an alert.
Use the filter "trashed" and delete trashed stages.

Make sure that the transitions to and from the deletes stages are deleted too.

### Expected result
All changes are correct, all alerts are correct

### Actual result
The alert for trashing the default stage is wrong (see  #21846)
Trashing several stages results in wrong conditions for stages. 

### Documentation Changes Required
no

### CODE REVIEW

File: administrator/components/com_workflow/Model/StageModel.php [hunk 1]
```diff
@@ -303,25 +304,120 @@ public function setDefault($pk, $value = 1)
 	 */
 	public function publish(&$pks, $value = 1)
 	{
-		$table = $this->getTable();
+		$db  = $this->getDbo();
+		$pks   = ArrayHelper::toInteger((array) $pks);
```
Comment (brianteeman) [inline administrator/components/com_workflow/Model/StageModel.php]: please align the = sign

File: administrator/components/com_workflow/Model/StageModel.php [hunk 2]
```diff
@@ -303,25 +304,120 @@ public function setDefault($pk, $value = 1)
...
+	 * @param   array    &$pks   A list of the primary keys to delete.
+	 *
+	 * @return  boolean  True on success.
+	 *
+	 * @since  __DEPLOY_VERSION__
+	 */
+	public function delete(&$pks)
+	{
+		$db  = $this->getDbo();
+		$pks   = ArrayHelper::toInteger((array) $pks);
```
Comment (brianteeman) [inline administrator/components/com_workflow/Model/StageModel.php]: please align the =

**Comment by wojsmol** [comment]
@chmst Please fix cs errors reported [here](http://ci.joomla.org/joomla/joomla-cms/11443/8).

File: administrator/components/com_workflow/Model/StageModel.php [hunk 3]
```diff
@@ -303,25 +304,124 @@ public function setDefault($pk, $value = 1)
...
+		if ($value !== 1)
+		{
+			// Clear pks 
+			self::checkDefaultStage($pks, $value);
+		}
+
+		if (!empty($pks))
+		{
+			// Change published for all stages
+			$return =  parent::publish($pks, $value);
```
Comment (Quy) [inline administrator/components/com_workflow/Model/StageModel.php]: Remove extra space after `=`

File: administrator/components/com_workflow/Model/StageModel.php [hunk 4]
```diff
@@ -303,25 +304,124 @@ public function setDefault($pk, $value = 1)
...
+
+			// If the stage is trashed, the transitions to and from this stage must be trashed too
+			if ($return && $value === -2)
+			{
+				$query = $db->getQuery(true)
+					->update($db->quoteName('#__workflow_transitions'))
+					->set($db->quoteName('published') . ' = 1')
+					->where(
+						$db->quoteName('from_stage_id') . ' IN (' . implode(',', $pks) . ') OR'
```
Comment (Quy) [inline administrator/components/com_workflow/Model/StageModel.php]: Add space after `OR`

File: administrator/components/com_workflow/Model/StageModel.php [hunk 5]
```diff
@@ -303,25 +304,124 @@ public function setDefault($pk, $value = 1)
...
+		$pks = ArrayHelper::toInteger((array) $pks);
+		$app = Factory::getApplication();
+
+		// Clear pks
+		self::checkDefaultStage($pks, -2);
+
+		if (!empty($pks))
+		{
+			// Change published for all stages where it is possible
+			$return =  parent::delete($pks);
```
Comment (Quy) [inline administrator/components/com_workflow/Model/StageModel.php]: Remove space after `=`

**Comment by laoneo** [comment]
Where do we stay here? Is the pr ready?

**Comment by chmst** [comment]
For me, it is ready, as it does not change something in the behaviour but fixes a but in a method call.

**Comment by wojsmol** [comment]
@chmst There is meagre conflict in `administrator/components/com_workflow/Model/StageModel.php`.

File: administrator/components/com_workflow/Model/StageModel.php [hunk 6]
```diff
@@ -303,25 +304,124 @@ public function setDefault($pk, $value = 1)
...
+			if ($value === -2)
 			{
-				if ($table->load(array('id' => $pk)) && $table->default)
+				$app->enqueueMessage(Text::_('COM_WORKFLOW_MSG_DELETE_DEFAULT'), 'error');
+			}
+			elseif ($value === 0)
+			{
+				$app->enqueueMessage(Text::_('COM_WORKFLOW_UNPUBLISH_DEFAULT_ERROR'), 'error');
+			}
+
```
Comment (Quy) [inline administrator/components/com_workflow/Model/StageModel.php]:
```suggestion

```

File: administrator/components/com_workflow/Model/WorkflowModel.php [hunk 7]
```diff
@@ -400,8 +400,18 @@ public function publish(&$pks, $value = 1)
 		{
 			if ($table->load($pk) && $value != 1 && $table->default)
 			{
+				$app = Factory::getApplication();
+
```
Comment (Quy) [inline administrator/components/com_workflow/Model/WorkflowModel.php]:
```suggestion

```

File: administrator/components/com_workflow/Model/WorkflowModel.php [hunk 8]
```diff
@@ -400,8 +400,18 @@ public function publish(&$pks, $value = 1)
...
+				
+				if ((int) $value === -2)
+				{
+					$app->enqueueMessage(Text::_('COM_WORKFLOW_MSG_DELETE_DEFAULT'), 'error');
+				}
+				elseif ((int) $value === 0)
+				{
+					$app->enqueueMessage(Text::_('COM_WORKFLOW_UNPUBLISH_DEFAULT_ERROR'), 'error');
+				}
+
```
Comment (Quy) [inline administrator/components/com_workflow/Model/WorkflowModel.php]:
```suggestion

```

**Comment by HRIT-Florian** [comment]
I have tested this item :white_check_mark: successfully on f0172022d09c927a6a1c9f8422ee316580ab364a<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/22594">issues.joomla.org/tracker/joomla-cms/22594</a>.</sub>

File: administrator/components/com_workflow/Model/StageModel.php [hunk 9]
```diff
@@ -12,9 +12,10 @@
 defined('_JEXEC') or die;
 
 use Joomla\CMS\Factory;
-use Joomla\CMS\Language\Text;
 use Joomla\CMS\MVC\Model\AdminModel;
+use Joomla\CMS\Language\Text;
```
Comment (Quy) [inline administrator/components/com_workflow/Model/StageModel.php]: Move back above line 15.

File: administrator/components/com_workflow/Model/StageModel.php [hunk 10]
```diff
@@ -303,25 +304,124 @@ public function setDefault($pk, $value = 1)
...
+						. $db->quoteName('to_stage_id') . ' IN (' . implode(',', $pks) . ')' 
+						);
+
+				$db->setQuery($query)->execute();
+			}
+		}
+
+		return $return;
+	}
+
```
Comment (Quy) [inline administrator/components/com_workflow/Model/StageModel.php]: Remove tab.

File: administrator/components/com_workflow/Model/StageModel.php [hunk 11]
```diff
@@ -303,25 +304,124 @@ public function setDefault($pk, $value = 1)
...
+				->delete($db->quoteName('#__workflow_transitions'))
+				->where($db->quoteName('to_stage_id') . ' = ' . (int) $pk, 'OR')
+				->whereIn($db->quoteName('from_stage_id'), $pks);
+
+			$db->setQuery($query)->execute();
+		}
+
+		return $return;
+	}
+
```
Comment (Quy) [inline administrator/components/com_workflow/Model/StageModel.php]: Remove tab.

File: administrator/components/com_workflow/Model/StageModel.php [hunk 12]
```diff
@@ -303,25 +304,124 @@ public function setDefault($pk, $value = 1)
...
-					$app->enqueueMessage(Text::_('COM_WORKFLOW_MSG_DELETE_DEFAULT'), 'error');
 					unset($pks[$i]);
 				}
 			}
 		}
-
-		return parent::publish($pks, $value);
 	}
 }
+
```
Comment (Quy) [inline administrator/components/com_workflow/Model/StageModel.php]: Remove line.

**Comment by roland-d** [comment]
@chmst Can you please fix the merge conflict so I can get this tested?

**Comment by chmst** [comment]
@roland-d such a long time since then - I will inspect this asap, then fix the conflicts

**Comment by vaibhavsTekdi** [comment]
I have tested this item :white_check_mark: successfully on c0196e0c6876029fdb1f081bb81009aab977e5fa<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/22594">issues.joomla.org/tracker/joomla-cms/22594</a>.</sub>

**Comment by pallaviu** [comment]
I have tested this item :white_check_mark: successfully on c0196e0c6876029fdb1f081bb81009aab977e5fa<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/22594">issues.joomla.org/tracker/joomla-cms/22594</a>.</sub>

File: administrator/components/com_workflow/Model/StageModel.php (hunk 12 above)
Comment (Quy) [inline administrator/components/com_workflow/Model/StageModel.php:427]:
```suggestion
```

File: administrator/components/com_workflow/Model/StageModel.php [hunk 13]
```diff
@@ -303,25 +304,124 @@ public function setDefault($pk, $value = 1)
...
 		$app = Factory::getApplication();
-		$extension = $app->getUserStateFromRequest('com_workflow.state.filter.extension', 'extension', null, 'cmd');
 
-		// Default item existence checks.
-		if ($value != 1)
+		// Check if one of the stages is the default stage
+		$query = $db->getQuery(true)
+			->select($db->quoteName(array('id')))
+			->from($db->quoteName('#__workflow_stages'))
+			->where($db->quoteName('default') . ' =1 ')
```
Comment (Quy) [inline administrator/components/com_workflow/Model/StageModel.php:400]:
```suggestion
			->where($db->quoteName('default') . ' = 1')
```

File: administrator/components/com_workflow/Model/StageModel.php (hunk 11 above)
Comment (Quy) [inline administrator/components/com_workflow/Model/StageModel.php:379]: Remove tab.

File: administrator/components/com_workflow/Model/StageModel.php [hunk 14]
```diff
@@ -303,25 +304,124 @@ public function setDefault($pk, $value = 1)
...
+	 * @param   integer  $value  The value of the published state.
+	 *
+	 * @return  void
+	 *
+	 * @since  __DEPLOY_VERSION__
+	 */
+	protected function checkDefaultStage(&$pks, $value = 1)
+	{
+		$db  = $this->getDbo();
 		$pks   = (array) $pks;
```
Comment (Quy) [inline administrator/components/com_workflow/Model/StageModel.php:393]:
```suggestion
		$pks = (array) $pks;
```

File: administrator/components/com_workflow/Model/StageModel.php [hunk 15]
```diff
@@ -303,25 +304,124 @@ public function setDefault($pk, $value = 1)
...
+			}
+			elseif ($value === 0)
+			{
+				$app->enqueueMessage(Text::_('COM_WORKFLOW_UNPUBLISH_DEFAULT_ERROR'), 'error');
+			}
+
+			// Remove the default stage from the list of keys
+			foreach ($pks as $i => $id)
+			{
+				if ((int) $default == (int) $id)
```
Comment (Quy) [inline administrator/components/com_workflow/Model/StageModel.php:419]:
```suggestion
				if ((int) $default === (int) $id)
```

**Comment by chmst** [comment]
Not longer valid, closed

**Pull request closed without merging**